fabric-model-events-api-v1/src/testmod/java/net/fabricmc/fabric/test/client/modelevents/ModelEventsTest.java

```diff
@@ -0,0 +1,190 @@
+import net.fabricmc.fabric.api.client.modelevents.v1.traversal.ModelVisitor;
+import net.fabricmc.fabric.impl.client.modelevents.ModelPartCallbacksImpl;
+import net.minecraft.block.entity.BlockEntityType;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.model.ModelPart;
+import net.minecraft.client.render.RenderLayer;
+import net.minecraft.client.render.WorldRenderer;
+import net.minecraft.client.render.entity.model.EntityModelPartNames;
+import net.minecraft.entity.EntityType;
+
+public final class ModelEventsTest implements ClientModInitializer {
+	private static final Logger LOGGER = LoggerFactory.getLogger(ModelEventsTest.class);
+
+	private static final ModelVisitor DEBUG_VISITOR = ModelVisitor.builder()
+            .visitCubes((matrices, cube) -> {
+                var debugConsumer = MinecraftClient.getInstance().getBufferBuilders().getEntityVertexConsumers().getBuffer(RenderLayer.getLines());
+                WorldRenderer.drawBox(matrices, debugConsumer, 0, 0, 0, cube.sizeX() / 16F, cube.sizeY() / 16F, cube.sizeZ() / 16F, 1, 1, 0, 1);
+                return true;
+            })
+            .visitFaces((matrices, face) -> {
+                var debugConsumer = MinecraftClient.getInstance().getBufferBuilders().getEntityVertexConsumers().getBuffer(RenderLayer.getLines());
+                matrices.push();
+
+                Vector3f center = face.center();
+                float cubeL = 0.01F;
+                matrices.translate(center.x / 16F, center.y / 16F, center.z / 16F);
+                Vector3f unitVector = face.direction().getUnitVector();
+                WorldRenderer.drawBox(matrices, debugConsumer, -cubeL, -cubeL, -cubeL, cubeL, cubeL, cubeL,
+                        0.5F + unitVector.x * 0.5F,
+                        0.5F + unitVector.y * 0.5F,
+                        0.5F + unitVector.z * 0.5F,
+                        1);
+                matrices.translate(unitVector.x / 16F, unitVector.y / 16F, unitVector.z / 16F);
+                WorldRenderer.drawBox(matrices, debugConsumer, -cubeL, -cubeL, -cubeL, cubeL, cubeL, cubeL,
+                        0.5F + unitVector.x * 0.5F,
+                        0.5F + unitVector.y * 0.5F,
+                        0.5F + unitVector.z * 0.5F,
+                        1);
+                matrices.pop();
+                return true;
+            }).build();
+
+	@Override
+	public void onInitializeClient() {
+	    List<AssertionError> errors = new ArrayList<>();
+
+	    test(errors, () -> {
+	        assert PartTreePath.of().isRoot();
+	    });
+	    test(errors, ModelEventsTest::checkEmptyPartPathCreation);
+	    test(errors, ModelEventsTest::checkPartPathParsing);
+	    test(errors, ModelEventsTest::checkPathIndexOfForShortPath);
+	    test(errors, ModelEventsTest::checkPathIndexOfForLongPath);
+	    test(errors, ModelEventsTest::checkPathComparisons);
+        test(errors, ModelEventsTest::registerEntityModelPartListener);
```

Comment:
This file is a bit weird with spaces and tabs. It's kind of hard to fix inside the IDE since it renders both the same and (confusingly) retains the mix of spaces and tabs when adding new lines.

I'll fix it later.